### ISSUE TITLE
GHCJS: run benchmark executables directly

### ISSUE THREAD

**Issue by harendra-kumar** [haskell/cabal#5831]
When building with GHCJS cabal is able to run the test executables but is not able to run the benchmark executables. I have added GHCJS support to `gauge` and we can now directly run benchmarks using `cabal new-bench`, only if cabal can support running them:

```
Linking /vol/hosts/cueball/workspace/packages/streamly/dist-newstyle/build/x86_64-linux/ghcjs-8.6.0.1/streamly-0.6.0/b/linear-async/build/linear-async/linear-async.jsexe (LinearOps,Main)
Running 1 benchmarks...
cabal: Error: Could not find benchmark program
"/vol/hosts/cueball/workspace/packages/streamly/dist-newstyle/build/x86_64-linux/ghcjs-8.6.0.1/streamly-0.6.0/b/linear-async/build/linear-async/linear-async".
Did you build the package first?
```

I can run manually using `node` though:

```
$ node /vol/hosts/cueball/workspace/packages/streamly/dist-newstyle/build/x86_64-linux/ghcjs-8.6.0.1/streamly-0.6.0/b/linear-async/build/linear-async/linear-async.jsexe/all.js

asyncly/unfoldr                          time                 138.4 μs  
asyncly/unfoldrM                         time                 58.00 ms  
asyncly/fromFoldable                     time                 2.357 ms 
asyncly/fromFoldableM                    time                 43.00 ms  
asyncly/foldMapWith                      time                 37.00 ms  
asyncly/foldMapWithM                     time                 34.00 ms  
...
```

**Comment by hvr** [comment]
@harendra-kumar the problem is that cabal currently only supports invoking executables which the OS can directly execute natively (i.e. where the kernel's executable loader is able to infer which method to invoke to run an executable passed to the `exec(3)` system call). We'd first have to introduce a new abstraction into Cabal which allows to execute artifacts which require some sort of interpreter (be it `node` or `jvm` or something else) sandwiched in.

**Comment by harendra-kumar** [comment]
But it does invoke the tests with `node`, so it seems to be already aware of the `node` interpreter:

```
$ cabal new-test test  --project-file cabal.project.ghcjs
...
Building test suite 'test' for streamly-0.6.0..
Linking /vol/hosts/cueball/workspace/packages/streamly/dist-newstyle/build/x86_64-linux/ghcjs-8.6.0.1/streamly-0.6.0/t/test/build/test/test.jsexe (Main)
Running 1 test suites...
Test suite test: RUNNING...
hello

Runners
  simple serially
  simple serially with IO
...
Finished in 15.9620 seconds
381 examples, 0 failures
Test suite test: PASS
Test suite logged to:
/vol/hosts/cueball/workspace/packages/streamly/dist-newstyle/build/x86_64-linux/ghcjs-8.6.0.1/streamly-0.6.0/t/test/test/streamly-0.6.0-test.log
```

We just need to do the same for benchmarks.

**Comment by hvr** [comment]
@harendra-kumar hrm... interesting... I think in the case of `cabal test` ghcjs might create some wrapper script which invokes the correct version of `node`....

**Comment by angerman** [comment]
"I have a patch!" (Although only for the generic test case in `ghc`.)

I should upstream [this](https://github.com/input-output-hk/cardano-sl/blob/develop/nix/patches/ghc/ghc-8.4.3-Cabal2201-allow-test-wrapper.patch) next week.

It could potentially be extended to work for benchmarks as well. The idea is that the end-user is responsible for providing the necessary shell script to to run/evaluate the executable.

**Comment by harendra-kumar** [comment]
Thanks @hvr for investigating this. @angerman that patch will provide great flexibility, it will be great to have the same for benchmarks as well. Though I do not have a good use case for this in my mind right now.

We still need to fix the problem that @hvr pointed out to enable running ghcjs benchmarks out of the box.

**Comment by hvr** [comment]
@harendra-kumar this has been fixed in master (see also #5832)

**Comment by harendra-kumar** [comment]
@hvr Awesome! Thanks!
